java/trace/src/main/scala/org/typelevel/otel4s/java/trace/TraceScope.scala

```diff
@@ -48,67 +49,84 @@ private[java] object TraceScope {
-              createScope(scopeRoot)
-
-            case Scope.Span(_, _, _) =>
-              createScope(scopeRoot)
-
-            case Scope.Noop =>
-              createScope(Scope.Noop)
-          }
-
-        def noopScope: F ~> F =
-          createScope(Scope.Noop)
-
-        private def createScope(scope: Scope): F ~> F =
-          new (F ~> F) {
-            def apply[A](fa: F[A]): F[A] =
-              MonadCancelThrow[F].bracket(local.getAndSet(scope).to[F])(
-                Function.const(fa)
-              )(p => local.set(p).to[F])
-          }
-
-        private def nextScope(scope: Scope, span: JSpan): Scope =
-          scope match {
-            case Scope.Root(ctx) =>
-              Scope.Span(
-                ctx.`with`(span),
-                span,
-                WrappedSpanContext(span.getSpanContext)
-              )
-
-            case Scope.Span(ctx, _, _) =>
-              Scope.Span(
-                ctx.`with`(span),
-                span,
-                WrappedSpanContext(span.getSpanContext)
-              )
-
-            case Scope.Noop =>
-              Scope.Noop
-          }
-
-      }
+    new TraceScope[F] {
+      val root: F[Scope.Root] =
+        L.applicative.pure(scopeRoot)
+
+      def current: F[Scope] =
+        L.ask[Scope]
+
+      def makeScope(span: JSpan): F[F ~> F] =
+        L.applicative.map(current)(scope => createScope(nextScope(scope, span)))
+
+      def rootScope: F[F ~> F] =
+        L.applicative.map(current) {
+          case Scope.Root(_) =>
+            createScope(scopeRoot)
+
+          case Scope.Span(_, _, _) =>
+            createScope(scopeRoot)
+
+          case Scope.Noop =>
+            createScope(Scope.Noop)
+        }
+
+      def noopScope: F ~> F =
+        createScope(Scope.Noop)
+
+      private def createScope(scope: Scope): F ~> F =
+        new (F ~> F) {
+          def apply[A](fa: F[A]): F[A] =
+            L.scope(fa)(scope)
+        }
+
+      private def nextScope(scope: Scope, span: JSpan): Scope =
+        scope match {
+          case Scope.Root(ctx) =>
+            Scope.Span(
+              ctx.`with`(span),
+              span,
+              WrappedSpanContext(span.getSpanContext)
+            )
+
+          case Scope.Span(ctx, _, _) =>
+            Scope.Span(
+              ctx.`with`(span),
+              span,
+              WrappedSpanContext(span.getSpanContext)
+            )
+
+          case Scope.Noop =>
+            Scope.Noop
+        }
     }
   }
 
+  private implicit def localForIoLocal[F[_]: MonadCancelThrow: LiftIO, E](
+      implicit ioLocal: IOLocal[E]
+  ): Local[F, E] =
+    new Local[F, E] {
+      def applicative =
+        Applicative[F]
+      def ask[E2 >: E] =
+        Functor[F].widen[E, E2](ioLocal.get.to[F])
+      def local[A](fa: F[A])(f: E => E): F[A] =
+        MonadCancelThrow[F].bracket(ioLocal.modify(e => (f(e), e)).to[F])(_ =>
+          fa
+        )(ioLocal.set(_).to[F])
+    }
```

Comment:
I guess this `LiftIO` is handy if you're lifting to a transformer and a going to be weird if you try to lift to a foreign base monad like Monix Task?  I'm not too certain about the whole `LiftIO` thing here.

Reply:
I'm not sure how one would even write such a `LiftIO` without a `Dispatcher` and an unsafe trip through `Future` delayed by the target monad.  Does anyone have an example of `LiftIO` for things other than transformers?

Reply:
I've thought about this before. Yeah, `LiftIO` to a foreign monad is probably not really an intended usecase.

> and an unsafe trip through `Future` delayed by the target monad

It might be interested to add `Dispatcher#unsafeToAsync` to skip the `Future` juggling.

Reply:
Skipping the future would be better.  Still, it might never evaluate the callback, won't interoperate with the foreign runtime aswon't rom the outcomes expressible in the callback, etc.  I wouldn't recommend it.

I don't think we can mapK a general Local[IO, E] to Local[F, E] without an F ~> IO.   We need an instance from IOLocal and LiftIO, or we need an abstraction over IOLocal.

Reply:
>  or we need an abstraction over IOLocal

Can you elaborate? I was thinking `Local` is that abstraction.

Reply:
No, I think you're right.

As of bc37ad5, we can trace from any Local.  The only `IOLocal` remaining is a convenience so nobody has to perform the ritual `IOLocal(default).map { implicit ioLocal => ... }` dance.